girder/models/file.py

```diff
@@ -94,18 +95,35 @@ def download(self, file, offset=0, headers=True, endByte=None,
         :type extraParameters: str or None
         """
         if file.get('assetstoreId'):
-            return self.getAssetstoreAdapter(file).downloadFile(
-                file, offset=offset, headers=headers, endByte=endByte,
-                contentDisposition=contentDisposition,
-                extraParameters=extraParameters)
+            try:
+                fileDownload = self.getAssetstoreAdapter(file).downloadFile(
+                    file, offset=offset, headers=headers, endByte=endByte,
+                    contentDisposition=contentDisposition,
+                    extraParameters=extraParameters)
+                if inspect.isgeneratorfunction(fileDownload) or inspect.isgenerator(fileDownload):
+                    def downloadGenerator():
+                        for data in fileDownload():
+                            yield data
+                        if endByte is None or endByte >= file['size']:
+                            events.trigger('download.complete', info={'file': file,
+                                                                      'redirect': False})
+                    return downloadGenerator
+                else:
+                    return fileDownload
```

Comment:
Ok, let's put a comment here explaining that fact (that this case should not be reached with any current assetstores).